### PR TITLE
ceph-objectstore-tool command to trim the pg log

### PR DESCRIPTION
**Pull request by dzafman**

### CODE REVIEW

File: src/tools/ceph_objectstore_tool.cc [hunk 1]
```diff
@@ -499,6 +499,107 @@ int write_pg(ObjectStore::Transaction &t, epoch_t epoch, pg_info_t &info,
...
+      }
+      if (debug) {
+	cerr << "read entry " << e << std::endl;
+      }
+      if (e.version.version > trim_to) {
+	done = true;
+	break;
+      }
+      keys_to_trim.insert(p->key());
+      new_tail = e.version;
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: I moved setting new_tail here so it is the version of the last element removed.  When new_tail == first log item's version then log_weirdness() complains:

 log_channel(cluster) log [ERR] : 1.0 log bound mismatch, info (tail,head] (12'22,12'41] actual [12'22,12'41]

See also http://tracker.ceph.com/issues/23269

**Comment by dzafman** [comment]
I need to build the luminous version and then I'l need to build the downstream version.

File: src/tools/ceph_objectstore_tool.cc (hunk 1 above)
Comment (jdurgin) [reply src/tools/ceph_objectstore_tool.cc]: ah, that looks right. I made an off-by-one error originally

File: qa/workunits/rados/repro_long_log.sh [hunk 2]
```diff
@@ -11,13 +13,16 @@ function test_log_size()
...
+sleep 5
+
 ceph osd pool create test 1 1 || true
 POOL_ID=$(ceph osd dump --format json | jq '.pools[] | select(.pool_name == "test") | .pool')
 PGID="${POOL_ID}.0"
 
-ceph tell osd.\* injectargs -- --osd-min-pg-log-entries 10
-ceph tell osd.\* injectargs -- --osd-max-pg-log-entries 20
+ceph tell osd.\* injectargs -- --osd-min-pg-log-entries 20
+ceph tell osd.\* injectargs -- --osd-max-pg-log-entries 30
```
[the remark this answers is not in the document]
Comment (jdurgin) [reply qa/workunits/rados/repro_long_log.sh]: that's odd, it was working on my luminous test - maybe slightly different trim behavior in master

**Comment by vumrao** [comment]
Tracker for this feature - http://tracker.ceph.com/issues/23242

**Comment by dzafman** [comment]
@jdurgin See new commit which will modify the database as each batch of keys is removed.  I added setting osd-pg-log-trim-max to the test and cout which shows key range being removed.

Test ../qa/run-standalone.sh repro_long_log.sh passes

../qa/standalone/osd/repro_long_log.sh:95: do_repro_long_log:  CEPH_ARGS='--osd-max-pg-log-entries=30 --osd-pg-log-trim-max=5'
../qa/standalone/osd/repro_long_log.sh:95: do_repro_long_log:  ceph-objectstore-tool --data-path td/repro-long-log/1 --pgid 1.0 --op trim-pg-log
Log bounds are: (0'0,16'41]
Removing keys 0000000016.00000000000000000001 - 0000000016.00000000000000000005
Removing keys 0000000016.00000000000000000006 - 0000000016.00000000000000000010
Removing keys 0000000016.00000000000000000011 - 0000000016.00000000000000000011
Finished trimming, now compacting...
Finished trimming pg log
...
ALL 1 TESTS PASSED

**Comment by dzafman** [comment]
retest this please

**Comment by dzafman** [comment]
Assuming osd_min_pg_log_entries is 3000, I think this snippet of bash script will do it.

```
sudo systemctl stop ceph-osd@_id_
for pg in $(ceph-objectstore-tool --data-path dev/osd0 --op list-pgs)
do
    CEPH_ARGS="--osd-max-pg-log-entries=3000" ceph-objectstore-tool --data-path _path_ --pgid $pg --op trim-pg-log
done
```

On the latest code add --no-mon-config to both ceph-objectstore-tool args if there is no mon running anywhere.

**Comment by charmingcow** [comment]
@dzafman
tks, did it can be directly used in version 10.2.10?
or replace code with 12.2.5 and recomplie in 10.2.10?

 ceph version 10.2.10 (5dc1e4c05cb68dbf62ae6fce3f0700e4654fdbbe)
 1: (()+0x92b54a) [0x7fbaf32b154a]
 2: (()+0xf370) [0x7fbaf130f370]
 3: (gsignal()+0x37) [0x7fbaef8d01d7]
 4: (abort()+0x148) [0x7fbaef8d18c8]
 5: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x7fbaf01d49d5]
 6: (()+0x5e946) [0x7fbaf01d2946]
 7: (()+0x5e973) [0x7fbaf01d2973]
 8: (()+0x5eb93) [0x7fbaf01d2b93]
 9: (ceph::buffer::list::iterator_impl<false>::copy(unsigned int, char*)+0xa5) [0x7fbaf33bc125]
 10: (pg_interval_t::decode(ceph::buffer::list::iterator&)+0x13b) [0x7fbaf2fb9b2b]
 11: (void decode<unsigned int, pg_interval_t>(std::map<unsigned int, pg_interval_t, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, pg_interval_t> > >&, ceph::buffer::list::iterator&)+0xff) [0x7fbaf2d7d51f]
 12: (PG::read_info(ObjectStore*, spg_t, coll_t const&, ceph::buffer::list&, pg_info_t&, std::map<unsigned int, pg_interval_t, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, pg_interval_t> > >&, unsigned char&)+0x278) [0x7fbaf2dde1d8]
 13: (PG::read_state(ObjectStore*, ceph::buffer::list&)+0x86) [0x7fbaf2dde896]
 14: (OSD::load_pgs()+0x9e2) [0x7fbaf2d231e2]
 15: (OSD::init()+0x212b) [0x7fbaf2d3577b]
 16: (main()+0x2c01) [0x7fbaf2c99131]
 17: (__libc_start_main()+0xf5) [0x7fbaef8bcb35]
 18: (()+0x35d4d7) [0x7fbaf2ce34d7]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

**Comment by dzafman** [comment]
@charmingcow You need to build with ceph/ceph#20882 which is targeted for 10.2.11.

**Pull request merged**